### PR TITLE
feat: allow for multiple Myokit simulations

### PR DESCRIPTION
**Pull request by eatyourgreens**
Refactor the Myokit model mixin to allow multiple simulations to be run. Each simulation has its own subject group, with subject dosing protocols and dosing events.

### CODE REVIEW

**Review by martinjrobins — CHANGES_REQUESTED** [review]
hi @eatyourgreens, this mostly looks right, but the number of simulations is not quite right:

You should perform 1 simulation for each distinct subject group. Each group of subjects will have 1 or more protocols associated with them (each dosing into a different variable). For each of those groups you want to gather the protocols for that group then run a simulation using all of the protocols for that group (you wont get an error since they will all be dosing into separate variables)

File: pkpdapp/pkpdapp/models/myokit_model_mixin.py [hunk 1]
```diff
@@ -45,97 +119,72 @@ def parse_mmt_string(mmt):
...
+        dataset=None,
+        variables=None,
+        time_max=None,
+        outputs=None
+    ):
+        override_tlag = self._get_override_tlag(variables),
+        if dataset is None:
+            subject_protocols = []
+        else:
+            subject_protocols = [
```
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply pkpdapp/pkpdapp/models/myokit_model_mixin.py]: I rewrote this so it's now expecting a list of subjects for each model, then getting the protocols from that list of subjects. My test groups only have one protocol per subject group, but it should work with more than one protocol.

File: pkpdapp/pkpdapp/models/myokit_model_mixin.py [hunk 2]
```diff
@@ -45,97 +119,72 @@ def parse_mmt_string(mmt):
...
+                override_tlag=override_tlag,
+                model=model,
+                time_max=time_max,
+                dosing_protocols=dosing_protocols
+            )
+            # TODO: take these from simulation model
+            sim.set_tolerance(abs_tol=1e-06, rel_tol=1e-08)
+            # Simulate, logging only state variables given by `outputs`
+            print('##########################################')
+            print(self.serialize_datalog(sim.run(time_max, log=outputs), model))
```
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply pkpdapp/pkpdapp/models/myokit_model_mixin.py]: All done now. THe API is sending back lists of simulations. Tests are updated, and the API schema. The Simulations tab simply loops through the response and generates a plot for each item. I'll work on the frontend in a separate PR. I think the Python changes are ready to go now (except for properly adding subject groups to the models.)

File: pkpdapp/pkpdapp/models/myokit_model_mixin.py [hunk 3]
```diff
@@ -689,3 +722,21 @@ def _get_dosing_events(
...
+        # frontend upload stepper
+        dataset_protocols = [
+            Protocol.objects.get(pk=p['protocol'])
+            for p in dataset.subjects.values('protocol').distinct()
+            if p['protocol'] is not None
+        ]
+        cohorts = [
+            dataset.subjects.filter(protocol=protocol)
+            for protocol in dataset_protocols
+        ]
```
Comment (eatyourgreens) [inline pkpdapp/pkpdapp/models/myokit_model_mixin.py:741]: Now that subject groups are in the app, this should ignore protocols and return the groups linked to this dataset's subjects

**Pull request merged**